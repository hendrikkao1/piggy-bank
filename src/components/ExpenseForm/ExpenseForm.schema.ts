import { z } from "zod";
import { CURRENCY } from "@/constants/currency";

export const ExpenseFormSchema = z.object({
  id: z.string(),
  amount: z
    .string()
    .min(1, { message: "amount.required" })
    .refine((val) => !Number.isNaN(parseFloat(val)), {
      message: "amount.number",
    })
    .refine((val) => parseFloat(val) > 0, {
      message: "amount.positive",
    }),
  currency: z.enum(CURRENCY, {
    errorMap: () => ({
      message: "currency.required",
    }),
  }),
  date: z
    .string()
    .min(1, { message: "date.required" })
    .regex(/\d\d\d\d-\d\d-\d\d/, { message: "date.pattern" }),
  recipient: z.string().min(1, { message: "recipient.required" }),
  type: z.string(),
});
