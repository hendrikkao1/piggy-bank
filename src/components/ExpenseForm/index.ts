export * from "./ExpenseForm.types";
export * from "./ExpenseFormFields";
