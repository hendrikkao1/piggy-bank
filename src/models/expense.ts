import { CURRENCY } from "@/constants/currency";
import { z } from "zod";
import Big from "big.js";

export const ExpenseSchema = z.object({
  amount: z.string().transform((val) => Big(val)),
  currency: z.enum(CURRENCY),
  date: z.string().transform((val) => new Date(val)),
  uuid: z.string().default(""),
  recipient: z.string(),
  type: z.string(),
});

export type IExpense = z.infer<typeof ExpenseSchema>;
