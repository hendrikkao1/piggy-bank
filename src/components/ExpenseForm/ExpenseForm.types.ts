import { z } from "zod";
import { ExpenseFormSchema } from "./ExpenseForm.schema";

export type IExpenseFormData = z.infer<typeof ExpenseFormSchema>;
