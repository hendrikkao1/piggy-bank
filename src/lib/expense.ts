import db, { IDBExpense } from "./db";
import { ExpenseSchema, IExpense } from "@/models/expense";

const expenseToDBExpense = (expense: IExpense): IDBExpense => {
  return {
    amount: expense.amount.toString(),
    currency: expense.currency,
    date: expense.date.toISOString(),
    uuid: expense.uuid,
    recipient: expense.recipient,
    type: expense.type,
  };
};

const addExpense = async (
  expense: IExpense,
): Promise<{
  data: IExpense | null;
  error: Error | null;
}> => {
  const { data, error } = await db.addExpense(expenseToDBExpense(expense));

  return {
    data: ExpenseSchema.parse(data),
    error: error ? new Error(error.message) : null,
  };
};

const getExpenseById = async (
  uuid: string,
): Promise<{
  data: IExpense | null;
  error: Error | null;
}> => {
  const { data, error } = await db.getExpenseById(uuid);

  return {
    data: ExpenseSchema.parse(data),
    error: error ? new Error(error.message) : null,
  };
};

const updateExpense = async (
  expense: IExpense,
): Promise<{
  data: IExpense | null;
  error: Error | null;
}> => {
  const { data, error } = await db.updateExpense(expenseToDBExpense(expense));

  return {
    data: ExpenseSchema.parse(data),
    error: error ? new Error(error.message) : null,
  };
};

const deleteExpenseById = async (
  uuid: string,
): Promise<{
  data: IExpense | null;
  error: Error | null;
}> => {
  const { data, error } = await db.deleteExpenseById(uuid);

  return {
    data: ExpenseSchema.parse(data),
    error: error ? new Error(error.message) : null,
  };
};

const getPaginatedExpenses = async (
  skip = 0,
  limit = Number.MAX_SAFE_INTEGER,
): Promise<{
  data: IExpense[] | null;
  error: Error | null;
}> => {
  const { data, error } = await db.getPaginatedExpenses(skip, limit);

  return {
    data: data ? data.map((d) => ExpenseSchema.parse(d)) : null,
    error: error ? new Error(error.message) : null,
  };
};

const expnese = {
  addExpense,
  deleteExpenseById,
  getExpenseById,
  getPaginatedExpenses,
  updateExpense,
};

export default expnese;
