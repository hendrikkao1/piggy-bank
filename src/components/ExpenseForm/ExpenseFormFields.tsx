import { CURRENCY } from "@/constants/currency";
import { IExpense } from "@/models/expense";
import { InputLabel } from "../InputLabel";
import { Input } from "../Input/Input";
import { Select } from "../Select/Select";

interface ExpenseFormFieldsProps {
  defaultValues: Partial<IExpense>;
}

export const ExpenseFormFields = ({
  defaultValues,
}: ExpenseFormFieldsProps) => (
  <div className="flex flex-col gap-6">
    <InputLabel label="recipient.label">
      <Input name="recipient" required defaultValue={defaultValues.recipient} />
    </InputLabel>
    <InputLabel label="date.label">
      <Input
        name="date"
        type="date"
        required
        defaultValue={defaultValues.date?.toISOString().split("T")[0]}
      />
    </InputLabel>
    <InputLabel label="amount.label">
      <Input
        name="amount"
        required
        type="number"
        defaultValue={defaultValues.amount?.toString()}
      />
    </InputLabel>
    <InputLabel label="currency.label">
      <Select name="currency" defaultValue={defaultValues.currency}>
        <option value="">pleaseChoose</option>
        {CURRENCY.map((currency) => (
          <option key={currency} value={currency}>
            {currency}
          </option>
        ))}
      </Select>
    </InputLabel>
    <InputLabel label={"type.label"}>
      <Input name="type" defaultValue={defaultValues.type} />
    </InputLabel>
  </div>
);
