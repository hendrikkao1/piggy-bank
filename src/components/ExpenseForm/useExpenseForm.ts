import { zodResolver } from "@hookform/resolvers/zod";
import Big from "big.js";
import { IExpense } from "@/models/expense";
import { ExpenseFormSchema } from "./ExpenseForm.schema";
import { SubmitHandler, useForm } from "react-hook-form";
import { IExpenseFormData } from "./ExpenseForm.types";

const mapExpenseToFormData = (
  expense?: Partial<IExpense>,
): Partial<IExpenseFormData> => ({
  ...expense,
  // As we don't register a field for id, we set it manually in case it is missing
  id: expense?.id ?? "",
  amount: expense?.amount?.toString(),
  date: expense?.date?.toISOString().split("T")[0],
});

const mapFormDataToExpense = (formData: IExpenseFormData): IExpense => ({
  ...formData,
  amount: Big(formData.amount),
  date: new Date(formData.date),
});

export interface UseExpenseFormOptions {
  defaultValue?: Partial<IExpense>;
  onSubmit: (expense: IExpense) => void;
}

export const useExpenseForm = ({
  defaultValue,
  onSubmit,
}: UseExpenseFormOptions) => {
  const formContext = useForm<IExpenseFormData>({
    defaultValues: mapExpenseToFormData(defaultValue),
    resolver: zodResolver(ExpenseFormSchema),
  });

  const handleSubmit: SubmitHandler<IExpenseFormData> = (formData) =>
    onSubmit(mapFormDataToExpense(formData));

  return {
    formContext,
    handleSubmit: formContext.handleSubmit(handleSubmit),
  };
};
