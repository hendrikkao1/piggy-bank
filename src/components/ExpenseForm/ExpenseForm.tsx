import { FormProvider } from "react-hook-form";
import { UseExpenseFormOptions, useExpenseForm } from "./useExpenseForm";
import React from "react";

type ExpeneseFormProps = React.PropsWithChildren<UseExpenseFormOptions>;

export const ExpeneseForm = ({ children, ...rest }: ExpeneseFormProps) => {
  const { formContext, handleSubmit } = useExpenseForm({
    defaultValue: rest.defaultValue,
    onSubmit: rest.onSubmit,
  });

  return (
    <FormProvider {...formContext}>
      <form noValidate onSubmit={handleSubmit}>
        {children}
      </form>
    </FormProvider>
  );
};
