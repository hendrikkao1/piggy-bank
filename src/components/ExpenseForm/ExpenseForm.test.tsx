import { fireEvent, render, waitFor } from "@/test-utils";
import { ExpeneseForm } from "./ExpenseForm";
import { IExpense } from "@/models/expense";
import Big from "big.js";

describe("ExpeneseForm", () => {
  it("renders", () => {
    const { container } = render(
      <ExpeneseForm onSubmit={jest.fn()}>
        <button>Submit</button>
      </ExpeneseForm>,
    );

    expect(container).toMatchSnapshot();
  });

  it("submits an expense", async () => {
    const mockOnSubmit = jest.fn();

    const mockExpense: IExpense = {
      amount: Big(100),
      currency: "EUR",
      date: new Date("2023-10-30T00:00:00.000Z"),
      id: "1",
      recipient: "John Doe",
      type: "food",
    };

    const screen = render(
      <ExpeneseForm defaultValue={mockExpense} onSubmit={mockOnSubmit}>
        <button>Submit</button>
      </ExpeneseForm>,
    );

    fireEvent.click(screen.getByRole("button", { name: "Submit" }));

    await waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalledWith(mockExpense);
    });
  });
});
