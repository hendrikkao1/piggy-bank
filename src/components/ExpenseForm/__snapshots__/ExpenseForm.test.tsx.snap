// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpeneseForm renders 1`] = `
<div>
  <form
    novalidate=""
  >
    <button>
      Submit
    </button>
  </form>
</div>
`;
